radixdlt-core/radixdlt/src/main/java/com/radixdlt/utils/streams/RoundRobinBackpressuredProcessor.java

```diff
@@ -0,0 +1,262 @@
+/*
+ * (C) Copyright 2021 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.utils.streams;
+
+import com.google.common.collect.Sets;
+import com.radixdlt.utils.Pair;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * A back-pressured processor that consumes upstream messages in a fair, round-robin manner.
+ * It is not strict, i.e. if there are messages on A, but no messages on B, then it will continue to consume
+ * messages from A (and will not block until any are available on B).
+ * The processor maintains a buffer of at most one message for each upstream publisher.
+ *
+ * Multiple downstream subscribers are allowed to connect to this processor.
+ * It doesn't maintain any cache for slower consumers though and, as a result, ALL downstream consumers
+ * need to signal a demand in order for message to be offered (to all of them, at the same time).
+ */
+public class RoundRobinBackpressuredProcessor<T> implements Publisher<T> {
```

Comment:
Missing final

Reply:
I'm not feeling good about actually implementing the Publisher ourselves: https://medium.com/@olehdokuka/mastering-own-reactive-streams-implementation-part-1-publisher-e8eaf928a78c
> Disclaimer: Rule number 1 of the Reactor club: don’t write your own Publisher. Even though the interface is simple, the set of rules about interactions between all these reactive streams interface is not. — Simon Baslé (Pivotal, Project Reactor).

For example, have we tested this Publisher against the Reactive Streams TCK: https://github.com/reactive-streams/reactive-streams-jvm/tree/master/tck?

If not, I would rather we try and use the available toolset to implement the Round-Robin functionality otherwise we may run into unexpected/weird errors down the line which are very hard to debug.